#! /bin/sh
patch -p1 -f $* < $0
exit $?

More intelligently populate the targ-include directory.

--- newlib-1.16.0/newlib/configure.in	2007-05-24 19:33:30.000000000 +0200
+++ newlib/newlib/configure.in	2008-01-22 15:27:33.000000000 +0100
@@ -127,6 +127,8 @@ AC_LIBTOOL_WIN32_DLL
 AM_PROG_LIBTOOL
 fi
 
+AC_PROG_LN_S
+
 AC_ARG_ENABLE(newlib_hw_fp,
 [  --enable-newlib-hw-fp    Turn on hardware floating point math],
 [case "${enableval}" in
@@ -141,7 +143,7 @@ AM_CONDITIONAL(NEWLIB_HW_FP, test x$newl
 LIBC_INCLUDE_GREP=`echo ${CC} | grep \/libc\/include`
 abs_newlib_basedir=`cd ${newlib_basedir} && pwd`
 if test -z "${LIBC_INCLUDE_GREP}"; then
-  CC_FOR_NEWLIB="${CC} -I$PWD/targ-include -I${abs_newlib_basedir}/libc/include"
+  CC_FOR_NEWLIB="${CC} -I$PWD/targ-include"
 else
   CC_FOR_NEWLIB="${CC}"
 fi
--- newlib-1.16.0/newlib/Makefile.am	2007-05-24 19:33:29.000000000 +0200
+++ newlib/newlib/Makefile.am	2008-01-22 15:36:40.000000000 +0100
@@ -195,52 +195,20 @@ all-recursive: stmp-targ-include
 stmp-targ-include: config.status
 	-rm -rf targ-include stmp-targ-include
 	mkdir targ-include targ-include/sys targ-include/machine targ-include/bits
-	cp newlib.h targ-include/newlib.h
-	-for i in $(srcdir)/libc/machine/$(machine_dir)/machine/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/machine/`basename $$i`; \
-	    else true; fi ; \
-	  done
-	-for i in $(srcdir)/libc/machine/$(machine_dir)/sys/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/sys/`basename $$i`; \
-	    else true; fi ; \
-	  done
-	-for i in $(srcdir)/libc/machine/$(machine_dir)/include/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/`basename $$i`; \
-	    else true; fi ; \
-	  done
-	-for i in $(srcdir)/libc/sys/$(sys_dir)/include/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/`basename $$i`; \
-	    else true; fi ; \
-	  done
-	-for i in $(srcdir)/libc/sys/$(sys_dir)/sys/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/sys/`basename $$i`; \
-	    else true; fi ; \
-	  done
-	-for i in $(srcdir)/libc/sys/$(sys_dir)/bits/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/bits/`basename $$i`; \
-	    else true; fi ; \
-	  done
-	-for i in $(srcdir)/libc/sys/$(sys_dir)/machine/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/machine/`basename $$i`; \
-	    else true; fi ; \
-	  done
-	-for i in $(srcdir)/libc/sys/$(sys_dir)/machine/$(machine_dir)/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/machine/`basename $$i`; \
-	    else true; fi ; \
-	  done
-	-for i in $(srcdir)/libc/sys/$(sys_dir)/machine/$(machine_dir)/include/*.h; do \
-	    if [ -f $$i ]; then \
-	      cp $$i targ-include/machine/`basename $$i`; \
-	    else true; fi ; \
-	  done
+	$(LN_S) $(abspath newlib.h) targ-include/newlib.h;
+	-for dir in sys/$(sys_dir)/machine/$(machine_dir) \
+		sys/$(sys_dir) machine/$(machine_dir) .; do \
+	  for include in include .; do \
+	    for subdir in machine sys bits .; do \
+	      for header in $(srcdir)/libc/$$dir/$$include/$$subdir/*.h; do \
+	        f=targ-include/$${subdir}/`basename $$header`; \
+	        if [ "x$${header##*/}" != x'*.h' -a ! -e "$$f" ]; then \
+	          $(LN_S) `realpath $$header` "$$f"; \
+	        fi; \
+	      done; \
+	    done; \
+	  done; \
+	done
 	touch $@
 
 CLEANFILES = targ-include stmp-targ-include
